h2o-core/src/main/java/hex/AUC2.java

```diff
@@ -325,16 +359,14 @@ public void reduce( AUCBuilder bldr ) {
       _n += bldr._n;
       //assert sorted();
 
-      // Merge elements with least squared-error increase until we get fewer
-      // than _nBins and no duplicates.  May require many merges.
-      while( _n > _nBins || dups() ) 
-        mergeOneBin();
+      // Merge duplicate rows in _ths.  May require many merges.  May or may  not cause reproducibility issue
+      removeDupsShrink(_workingNBins, true);
     }
 
-    private void mergeOneBin( ) {
+    private int mergeOneBin( ) {
       // Too many bins; must merge bins.  Merge into bins with least total
       // squared error.  Horrible slowness linear arraycopy.
-      int ssx = find_smallest();
+      int ssx = (_ssx > 0)? _ssx : find_smallest(); // Dups() will set _ssx
```

Comment:
@wendycwong can you please explain this line? find_smallest already internally uses the cache (ssx), this is just duplicating what is in find_smallest, isn't it?

Reply:
What I am doing here is when we found duplicates, we know index already.  There is no need to repeat and find find the smallest index again.

Reply:
This is true @wendycwong, however, find_smallest already does the exact same thing internally for you.

Perhaps the issue is that you are running with enabled assertions and find_smallest calls find_smallest_impl in an assertion. I believe you can remote the check from here and remove the assertion from find_smallest. This is the right fix in my opinion.